### PR TITLE
Add registration/icp

### PR DESCRIPTION
**Pull request by at-wat**
Add MVP implementation of 3D/6DoF ICP registration.

### TODO in this PR
- [x] point-to-point correspondence estimation
- [x] objective function (residual vector norm)
- [x] minimization by gradient

### Not to do in this PR
- other correspondence model (line-to-point/surface-to-point)
- minimization using hessian (Newton's method)

### CODE REVIEW

File: pc/minmax.go [hunk 1]
```diff
@@ -30,3 +30,23 @@ func MinMaxVec3(pp *PointCloud) (mat.Vec3, mat.Vec3, error) {
 	}
 	return min, max, nil
 }
+
+func MinMaxVec3RandomAccessor(ra Vec3RandomAccessor) (mat.Vec3, mat.Vec3, error) {
```
Comment (at-wat) [inline pc/minmax.go:34]: Will be replaced by #13

**Comment by at-wat** [comment]
@f-fl0 please take a look

File: pc/registration/icp/icp.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/seqsense/pcgol/pc/storage"
+
+	"github.com/seqsense/pcgol/internal/gnuplot"
+)
+
+var (
+	ErrNeedGradient = errors.New("need gradient output of Evaluator")
+)
+
+var DefaultGradientWeight = mat.Vec6{0.3, 0.3, 0.3, 0.3, 0.3, 0.3}
```
Comment (f-fl0) [inline pc/registration/icp/icp.go:18]: My understanding of the weights was that it is a way to "bias" the transformation towards some dimensions but if they are all the same, it is pretty much a scaling factor. As the delta transformation is already scaled by `factorIter` at each iteration, why not defaulting to `1`?

**Review by f-fl0 — COMMENTED** [review]
As the `gnuplot` stuff are pretty independent from `pcgol` do you think it could be worth having a separate repository for it instead of storing the code in here?

File: pc/registration/icp/rodrigues.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+		0, 1, 0, 0,
+		0, 0, 1, 0,
+		0, 0, 0, 1,
+	}
+	var f0, f1 float32
+	if ang < 0.1 {
+		f0, f1 = 1, 0.5
+	} else {
+		f0 = float32(math.Sin(float64(ang))) / ang
+		f1 = float32(1-math.Cos(float64(ang))) / (ang * ang)
```
[the remark this answers is not in the document]
Comment (at-wat) [reply pc/registration/icp/rodrigues.go:30]: It's true when the input vector is normalized.
The length of the input vector may be close to zero and normalization of the input vector causes division by zero problem. This code avoids zero division by using approximation that `sin(ang) / ang ~= 1` and `(1-cos(ang))/ang^2 = 1/2` when `ang ~= 0`.

File: mat/mat4_test.go [hunk 4]
```diff
@@ -140,3 +185,34 @@ func TestMat4_Inv(t *testing.T) {
...
+		3, 7, 11, 15,
+		4, 8, 12, 16,
+	}
+	trans := m.Transpose()
+	if !reflect.DeepEqual(expected, trans) {
+		t.Errorf("Expected %v, got %v", expected, trans)
+	}
+}
+
+func ExampleMat4_String() {
```
[the remark this answers is not in the document]
Comment (f-fl0) [reply mat/mat4_test.go:208]: My bad I didn't know that.

File: pc/registration/icp/icp.go (hunk 2 above)
Comment (at-wat) [reply pc/registration/icp/icp.go:18]: When I set it to 1, it often overshoots and not converged well.
In gradient descent method, factor should be just `small enough`. https://en.wikipedia.org/wiki/Gradient_descent

File: pc/registration/icp/evaluator.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+		)
+		time.Sleep(debugPlotInterval)
+	}
+
+	f := 1 / float32(len(pairs))
+	out.Value *= f
+	for i := 0; i < 3; i++ {
+		out.Gradient[i] *= f
+	}
+	fn := 1 / float32(num)
```
[the remark this answers is not in the document]
Comment (at-wat) [reply pc/registration/icp/evaluator.go]: The result is same but reordered the check. 
adcf072

**Pull request merged**